File: apps/api/internal/service/travel_allowance_preview.go
package service

import (
	"context"
	"errors"
	"time"

	"github.com/google/uuid"
	"github.com/shopspring/decimal"

	"github.com/tolga/terp/internal/calculation"
	"github.com/tolga/terp/internal/model"
)

var (
	ErrTravelPreviewRuleSetNotFound   = errors.New("rule set not found for travel allowance preview")
	ErrTravelPreviewRuleSetIDRequired = errors.New("rule set ID is required for preview")
	ErrTravelPreviewTripTypeRequired  = errors.New("trip type is required (local or extended)")
	ErrTravelPreviewInvalidTripType   = errors.New("trip type must be 'local' or 'extended'")
	ErrTravelPreviewDistanceRequired  = errors.New("distance is required for local travel preview")
	ErrTravelPreviewDurationRequired  = errors.New("duration is required for local travel preview")
	ErrTravelPreviewDatesRequired     = errors.New("start_date and end_date are required for extended travel preview")
	ErrTravelPreviewNoMatchingRule    = errors.New("no matching local travel rule found for given distance and duration")
	ErrTravelPreviewNoExtendedRule    = errors.New("no active extended travel rule found for this rule set")
)

// previewRuleSetRepository defines the interface for rule set lookup.
type previewRuleSetRepository interface {
	GetByID(ctx context.Context, id uuid.UUID) (*model.TravelAllowanceRuleSet, error)
}

// previewLocalRuleRepository defines the interface for local rule lookup.
type previewLocalRuleRepository interface {
	ListByRuleSet(ctx context.Context, ruleSetID uuid.UUID) ([]model.LocalTravelRule, error)
}

// previewExtendedRuleRepository defines the interface for extended rule lookup.
type previewExtendedRuleRepository interface {
	ListByRuleSet(ctx context.Context, ruleSetID uuid.UUID) ([]model.ExtendedTravelRule, error)
}

// TravelAllowancePreviewService handles travel allowance preview calculations.
type TravelAllowancePreviewService struct {
	ruleSetRepo      previewRuleSetRepository
	localRuleRepo    previewLocalRuleRepository
	extendedRuleRepo previewExtendedRuleRepository
}

// NewTravelAllowancePreviewService creates a new TravelAllowancePreviewService.
func NewTravelAllowancePreviewService(
	ruleSetRepo previewRuleSetRepository,
	localRuleRepo previewLocalRuleRepository,
	extendedRuleRepo previewExtendedRuleRepository,
) *TravelAllowancePreviewService {
	return &TravelAllowancePreviewService{
		ruleSetRepo:      ruleSetRepo,
		localRuleRepo:    localRuleRepo,
		extendedRuleRepo: extendedRuleRepo,
	}
}

// TravelAllowancePreviewInput holds input for a travel allowance preview.
type TravelAllowancePreviewInput struct {
	RuleSetID        uuid.UUID
	TripType         string // "local" or "extended"
	DistanceKm       float64
	DurationMinutes  int
	StartDate        time.Time
	EndDate          time.Time
	ThreeMonthActive bool
}

// TravelAllowancePreviewResult holds the preview result.
type TravelAllowancePreviewResult struct {
	TripType       string
	RuleSetID      uuid.UUID
	RuleSetName    string
	TaxFreeTotal   decimal.Decimal
	TaxableTotal   decimal.Decimal
	TotalAllowance decimal.Decimal
	Breakdown      []TravelAllowanceBreakdownItem
}

// TravelAllowanceBreakdownItem describes a single line in the breakdown.
type TravelAllowanceBreakdownItem struct {
	Description     string
	Days            int
	TaxFreeAmount   decimal.Decimal
	TaxableAmount   decimal.Decimal
	TaxFreeSubtotal decimal.Decimal
	TaxableSubtotal decimal.Decimal
}

// Preview calculates a travel allowance preview.
func (s *TravelAllowancePreviewService) Preview(ctx context.Context, input TravelAllowancePreviewInput) (*TravelAllowancePreviewResult, error) {
	// Validate input
	if input.RuleSetID == uuid.Nil {
		return nil, ErrTravelPreviewRuleSetIDRequired
	}
	if input.TripType == "" {
		return nil, ErrTravelPreviewTripTypeRequired
	}
	if input.TripType != "local" && input.TripType != "extended" {
		return nil, ErrTravelPreviewInvalidTripType
	}

	// Fetch rule set
	ruleSet, err := s.ruleSetRepo.GetByID(ctx, input.RuleSetID)
	if err != nil {
		return nil, ErrTravelPreviewRuleSetNotFound
	}

	if input.TripType == "local" {
		return s.previewLocal(ctx, input, ruleSet)
	}
	return s.previewExtended(ctx, input, ruleSet)
}

func (s *TravelAllowancePreviewService) previewLocal(ctx context.Context, input TravelAllowancePreviewInput, ruleSet *model.TravelAllowanceRuleSet) (*TravelAllowancePreviewResult, error) {
	if input.DistanceKm <= 0 && input.DurationMinutes <= 0 {
		return nil, ErrTravelPreviewDistanceRequired
	}

	// Fetch local rules for this rule set
	rules, err := s.localRuleRepo.ListByRuleSet(ctx, input.RuleSetID)
	if err != nil {
		return nil, err
	}

	// Build calculation input
	calcRules := make([]calculation.LocalTravelRuleInput, 0, len(rules))
	for _, r := range rules {
		if !r.IsActive {
			continue
		}
		calcRules = append(calcRules, calculation.LocalTravelRuleInput{
			MinDistanceKm:      r.MinDistanceKm,
			MaxDistanceKm:      r.MaxDistanceKm,
			MinDurationMinutes: r.MinDurationMinutes,
			MaxDurationMinutes: r.MaxDurationMinutes,
			TaxFreeAmount:      r.TaxFreeAmount,
			TaxableAmount:      r.TaxableAmount,
		})
	}

	calcInput := calculation.LocalTravelInput{
		DistanceKm:      decimal.NewFromFloat(input.DistanceKm),
		DurationMinutes: input.DurationMinutes,
		Rules:           calcRules,
	}

	calcOutput := calculation.CalculateLocalTravelAllowance(calcInput)

	if !calcOutput.Matched {
		return nil, ErrTravelPreviewNoMatchingRule
	}

	result := &TravelAllowancePreviewResult{
		TripType:       "local",
		RuleSetID:      ruleSet.ID,
		RuleSetName:    ruleSet.Name,
		TaxFreeTotal:   calcOutput.TaxFreeTotal,
		TaxableTotal:   calcOutput.TaxableTotal,
		TotalAllowance: calcOutput.TotalAllowance,
		Breakdown: []TravelAllowanceBreakdownItem{
			{
				Description:     "Local travel allowance",
				Days:            1,
				TaxFreeAmount:   calcOutput.TaxFreeTotal,
				TaxableAmount:   calcOutput.TaxableTotal,
				TaxFreeSubtotal: calcOutput.TaxFreeTotal,
				TaxableSubtotal: calcOutput.TaxableTotal,
			},
		},
	}

	return result, nil
}

func (s *TravelAllowancePreviewService) previewExtended(ctx context.Context, input TravelAllowancePreviewInput, ruleSet *model.TravelAllowanceRuleSet) (*TravelAllowancePreviewResult, error) {
	if input.StartDate.IsZero() || input.EndDate.IsZero() {
		return nil, ErrTravelPreviewDatesRequired
	}

	// Fetch extended rules for this rule set
	rules, err := s.extendedRuleRepo.ListByRuleSet(ctx, input.RuleSetID)
	if err != nil {
		return nil, err
	}

	// Find first active rule
	var activeRule *model.ExtendedTravelRule
	for i := range rules {
		if rules[i].IsActive {
			activeRule = &rules[i]
			break
		}
	}
	if activeRule == nil {
		return nil, ErrTravelPreviewNoExtendedRule
	}

	calcInput := calculation.ExtendedTravelInput{
		StartDate:        input.StartDate,
		EndDate:          input.EndDate,
		ThreeMonthActive: input.ThreeMonthActive,
		Rule: calculation.ExtendedTravelRuleInput{
			ArrivalDayTaxFree:      activeRule.ArrivalDayTaxFree,
			ArrivalDayTaxable:      activeRule.ArrivalDayTaxable,
			DepartureDayTaxFree:    activeRule.DepartureDayTaxFree,
			DepartureDayTaxable:    activeRule.DepartureDayTaxable,
			IntermediateDayTaxFree: activeRule.IntermediateDayTaxFree,
			IntermediateDayTaxable: activeRule.IntermediateDayTaxable,
			ThreeMonthEnabled:      activeRule.ThreeMonthEnabled,
			ThreeMonthTaxFree:      activeRule.ThreeMonthTaxFree,
			ThreeMonthTaxable:      activeRule.ThreeMonthTaxable,
		},
	}

	calcOutput := calculation.CalculateExtendedTravelAllowance(calcInput)

	breakdown := make([]TravelAllowanceBreakdownItem, 0, len(calcOutput.Breakdown))
	for _, item := range calcOutput.Breakdown {
		breakdown = append(breakdown, TravelAllowanceBreakdownItem{
			Description:     item.Description,
			Days:            item.Days,
			TaxFreeAmount:   item.TaxFreeAmount,
			TaxableAmount:   item.TaxableAmount,
			TaxFreeSubtotal: item.TaxFreeSubtotal,
			TaxableSubtotal: item.TaxableSubtotal,
		})
	}

	result := &TravelAllowancePreviewResult{
		TripType:       "extended",
		RuleSetID:      ruleSet.ID,
		RuleSetName:    ruleSet.Name,
		TaxFreeTotal:   calcOutput.TaxFreeTotal,
		TaxableTotal:   calcOutput.TaxableTotal,
		TotalAllowance: calcOutput.TotalAllowance,
		Breakdown:      breakdown,
	}

	return result, nil
}

File: src/server/routers/travelAllowancePreview.ts
/**
 * Travel Allowance Preview Router
 *
 * Provides travel allowance calculation preview via tRPC.
 *
 * Replaces the Go backend travel allowance preview endpoint:
 * - POST /travel-allowance-preview -> travelAllowancePreview.preview (query)
 *
 * @see apps/api/internal/service/travel_allowance_preview.go
 * @see apps/api/internal/calculation/travel_allowance.go
 */
import { z } from "zod"
import { TRPCError } from "@trpc/server"
import { createTRPCRouter, tenantProcedure } from "@/trpc/init"
import { requirePermission } from "../middleware/authorization"
import { permissionIdByKey } from "../lib/permission-catalog"
import {
  calculateLocalTravelAllowance,
  calculateExtendedTravelAllowance,
} from "@/lib/calculation/travel-allowance"
import type {
  LocalTravelRuleInput,
  ExtendedTravelRuleInput,
} from "@/lib/calculation/travel-allowance"

// --- Permission Constants ---

const TRAVEL_ALLOWANCE_MANAGE = permissionIdByKey("travel_allowance.manage")!

// --- Output Schemas ---

const breakdownItemSchema = z.object({
  description: z.string(),
  days: z.number(),
  taxFreeAmount: z.number(),
  taxableAmount: z.number(),
  taxFreeSubtotal: z.number(),
  taxableSubtotal: z.number(),
})

const previewOutputSchema = z.object({
  tripType: z.string(),
  ruleSetId: z.string().uuid(),
  ruleSetName: z.string(),
  taxFreeTotal: z.number(),
  taxableTotal: z.number(),
  totalAllowance: z.number(),
  breakdown: z.array(breakdownItemSchema),
})

// --- Input Schemas ---

const previewInputSchema = z.object({
  ruleSetId: z.string().uuid(),
  tripType: z.enum(["local", "extended"]),
  distanceKm: z.number().optional().default(0),
  durationMinutes: z.number().int().optional().default(0),
  startDate: z.string().optional(),
  endDate: z.string().optional(),
  threeMonthActive: z.boolean().optional().default(false),
})

// --- Router ---

export const travelAllowancePreviewRouter = createTRPCRouter({
  /**
   * travelAllowancePreview.preview -- Calculates a travel allowance preview.
   *
   * For local trips: matches distance/duration against active local rules.
   * For extended trips: computes day breakdown using active extended rule.
   *
   * Requires: travel_allowance.manage permission
   */
  preview: tenantProcedure
    .use(requirePermission(TRAVEL_ALLOWANCE_MANAGE))
    .input(previewInputSchema)
    .output(previewOutputSchema)
    .query(async ({ ctx, input }) => {
      const tenantId = ctx.tenantId!

      // Fetch rule set
      const ruleSet = await ctx.prisma.travelAllowanceRuleSet.findFirst({
        where: { id: input.ruleSetId, tenantId },
      })
      if (!ruleSet) {
        throw new TRPCError({
          code: "NOT_FOUND",
          message: "Rule set not found",
        })
      }

      if (input.tripType === "local") {
        return previewLocal(ctx, input, ruleSet)
      }
      return previewExtended(ctx, input, ruleSet)
    }),
})

// --- Private helpers ---

async function previewLocal(
  ctx: { prisma: { localTravelRule: { findMany: Function } } },
  input: {
    ruleSetId: string
    distanceKm: number
    durationMinutes: number
  },
  ruleSet: { id: string; name: string }
) {
  if (input.distanceKm <= 0 && input.durationMinutes <= 0) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message:
        "Distance or duration is required for local travel preview",
    })
  }

  // Fetch local rules for this rule set (active only)
  const rules = await ctx.prisma.localTravelRule.findMany({
    where: { ruleSetId: input.ruleSetId, isActive: true },
    orderBy: [{ sortOrder: "asc" }, { minDistanceKm: "asc" }],
  })

  // Build calculation input (convert Prisma Decimals to numbers)
  const calcRules: LocalTravelRuleInput[] = rules.map(
    (r: Record<string, unknown>) => ({
      minDistanceKm: Number(r.minDistanceKm),
      maxDistanceKm: r.maxDistanceKm != null ? Number(r.maxDistanceKm) : null,
      minDurationMinutes: Number(r.minDurationMinutes),
      maxDurationMinutes:
        r.maxDurationMinutes != null ? Number(r.maxDurationMinutes) : null,
      taxFreeAmount: Number(r.taxFreeAmount),
      taxableAmount: Number(r.taxableAmount),
    })
  )

  const calcOutput = calculateLocalTravelAllowance({
    distanceKm: input.distanceKm,
    durationMinutes: input.durationMinutes,
    rules: calcRules,
  })

  if (!calcOutput.matched) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: "No matching local travel rule found",
    })
  }

  return {
    tripType: "local" as const,
    ruleSetId: ruleSet.id,
    ruleSetName: ruleSet.name,
    taxFreeTotal: calcOutput.taxFreeTotal,
    taxableTotal: calcOutput.taxableTotal,
    totalAllowance: calcOutput.totalAllowance,
    breakdown: [
      {
        description: "Local travel allowance",
        days: 1,
        taxFreeAmount: calcOutput.taxFreeTotal,
        taxableAmount: calcOutput.taxableTotal,
        taxFreeSubtotal: calcOutput.taxFreeTotal,
        taxableSubtotal: calcOutput.taxableTotal,
      },
    ],
  }
}

async function previewExtended(
  ctx: { prisma: { extendedTravelRule: { findMany: Function } } },
  input: {
    ruleSetId: string
    startDate?: string
    endDate?: string
    threeMonthActive: boolean
  },
  ruleSet: { id: string; name: string }
) {
  if (!input.startDate || !input.endDate) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message:
        "Start date and end date are required for extended travel preview",
    })
  }

  const startDate = new Date(input.startDate + "T00:00:00.000Z")
  const endDate = new Date(input.endDate + "T00:00:00.000Z")

  if (isNaN(startDate.getTime()) || isNaN(endDate.getTime())) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message:
        "Start date and end date are required for extended travel preview",
    })
  }

  // Fetch extended rules for this rule set (active only)
  const rules = await ctx.prisma.extendedTravelRule.findMany({
    where: { ruleSetId: input.ruleSetId, isActive: true },
    orderBy: [{ sortOrder: "asc" }],
  })

  // Find first active rule
  if (rules.length === 0) {
    throw new TRPCError({
      code: "BAD_REQUEST",
      message: "No active extended travel rule found for this rule set",
    })
  }

  const activeRule = rules[0] as Record<string, unknown>

  const extRule: ExtendedTravelRuleInput = {
    arrivalDayTaxFree: Number(activeRule.arrivalDayTaxFree),
    arrivalDayTaxable: Number(activeRule.arrivalDayTaxable),
    departureDayTaxFree: Number(activeRule.departureDayTaxFree),
    departureDayTaxable: Number(activeRule.departureDayTaxable),
    intermediateDayTaxFree: Number(activeRule.intermediateDayTaxFree),
    intermediateDayTaxable: Number(activeRule.intermediateDayTaxable),
    threeMonthEnabled: activeRule.threeMonthEnabled as boolean,
    threeMonthTaxFree: Number(activeRule.threeMonthTaxFree),
    threeMonthTaxable: Number(activeRule.threeMonthTaxable),
  }

  const calcOutput = calculateExtendedTravelAllowance({
    startDate,
    endDate,
    threeMonthActive: input.threeMonthActive,
    rule: extRule,
  })

  return {
    tripType: "extended" as const,
    ruleSetId: ruleSet.id,
    ruleSetName: ruleSet.name,
    taxFreeTotal: calcOutput.taxFreeTotal,
    taxableTotal: calcOutput.taxableTotal,
    totalAllowance: calcOutput.totalAllowance,
    breakdown: calcOutput.breakdown.map((item) => ({
      description: item.description,
      days: item.days,
      taxFreeAmount: item.taxFreeAmount,
      taxableAmount: item.taxableAmount,
      taxFreeSubtotal: item.taxFreeSubtotal,
      taxableSubtotal: item.taxableSubtotal,
    })),
  }
}
